### ISSUE TITLE
hi，i want to apply nash-mtl to my multi-task learning net.

### ISSUE THREAD

**Issue by yushengjiexy** [AvivNavon/nash-mtl#13]
Hello, my network is a 2-task network based on ResNet-18. I have tried PCGrad, am currently trying nash-mtl. But I found that the multi-task training effect decreased with the addition of PCGrad, and the network with nash-mtl could not even converge. May I ask if there are any applicable conditions for nash-mtl? For example, what are the requirements for the loss of each task, what are the requirements for multi-task network, and other applicable conditions? In addition, how to adjust the hyperparameter of nash-mtl, 'update_weights_every', "optim_niter", 'max_norm'?

**Comment by yushengjiexy** [comment]
The loss  of my two tasks are approximately 600 and 700. When calculating the weight of losses in Nashmtl,  the weight always was  about 10-3, and the total loss of the network became about 0.5, which caused my network to be unable to converge. How to change the hyperparameter?

**Comment by AvivNavon** [comment]
Hi @yushengjiexy,
Did you try changing the learning rate?

Regarding the specific HPs in your original question:
-  `update_weights_every`: keep it 1. This is the frequency in which the task weights are updated. 
- `optim_niter`: no need to change this. 
- `max_norm`: this is the maximal norm of the resulting gradient.

**Comment by yushengjiexy** [comment]
Thank you for your reply!

I have not  tried changing the learning rate.I found that there are the gradient of some parameters in my network are high. Is this the reason why the alpha parameter is too small?I add some print to my code, and here is what I want to show you:

![image](https://github.com/AvivNavon/nash-mtl/assets/63044657/65124e94-bc50-4880-be26-047a948f070a)

**Comment by yushengjiexy** [comment]
By the way, I am using nashmtl integrated into libmtl:https://github.com/median-research-group/LibMTL
I also tried your official code, but there is also this issue.

**Comment by AvivNavon** [comment]
The norm of the update direction obtained by NashMTL is constant, so it is OK that the task weights are small. Could you share some learning curves? I suggest you try to modify the lr.

**Comment by yushengjiexy** [comment]
hi, this is my learning curves.
 x: epochs, y: learning rate(1e-5)
![image](https://github.com/AvivNavon/nash-mtl/assets/63044657/6fcc892a-7929-4a10-9421-b361e938f2bc)

**Comment by AvivNavon** [comment]
Hi @yushengjiexy, thanks. I meant the curves of the loss during training.
DIt you tried increasing the initial learning rate (e.g. from 1e-5 to 1e-4)?

**Comment by yushengjiexy** [comment]
hi, @AvivNavon, sorry, my English is poor.
1.x : epoch, y : loss, with nashmtl
![image](https://github.com/AvivNavon/nash-mtl/assets/63044657/51cc9162-77d0-4dc6-aa8e-ae0bb0efd258)

x : epoch, y : loss, without nashmtl
![image](https://github.com/AvivNavon/nash-mtl/assets/63044657/c409bc61-f91f-4dd4-b62a-2f2747137d5e)

2.sorry, I did not increasing the initial learning rate.

**Comment by AvivNavon** [comment]
Thank you. Could you please try modifying the lr?

**Comment by yushengjiexy** [comment]
Because I print the loss for the 40th iteration of each epoch, the corresponding loss for each epoch I draw here is actually the loss for the 40th iteration of this epoch.

**Comment by yushengjiexy** [comment]
> Thank you. Could you please try modifying the lr?

ok，i will try increasing the initial learning rate from 1e-5 to 1e-4. I will tell you the result.
Thank you!

**Comment by yushengjiexy** [comment]
@AvivNavon 
hi, i change the initial learning rate to 1e-3. But it don't work.
![image](https://github.com/AvivNavon/nash-mtl/assets/63044657/bbe05e7f-e421-4baa-9e51-389ed0ee499c)

**Comment by AvivNavon** [comment]
Thanks. Are these the raw losses or weighted losses? 
I see that loss2 is higher here, and above it was loss1, is this a mistake?

**Comment by yushengjiexy** [comment]
> 

sorry, this is a mistake.

**Comment by yushengjiexy** [comment]
> Thanks. Are these the raw losses or weighted losses? I see that loss2 is higher here, and above it was loss1, is this a mistake?

1.these are raw losses.
2.sorry, this is a mistake.

**Comment by AvivNavon** [comment]
Thanks. 
It's hard to debug without access to the code and data.
Maybe look at the norm of the shared param gradients before we clip it here:
https://github.com/AvivNavon/nash-mtl/blob/main/methods/weight_methods.py#L268

**Comment by yushengjiexy** [comment]
> Thanks. It's hard to debug without access to the code and data. Maybe look at the norm of the shared param gradients before we clip it here: https://github.com/AvivNavon/nash-mtl/blob/main/methods/weight_methods.py#L268

1.sorry, since I am an employeer, I cann't share the code and data with you。
2.I try to look at the norm of  the shared param gradients,  i found the gradients are very small. In shallow networks, there is no gradient in bn.weight. I think there is a problem of gradient disappearance.

**Comment by AvivNavon** [comment]
Thanks for checking. As I said, it's hard to debug with no access to the network and dataset. 
Nash-MTL is scale-invariant, so loss scale changes between tasks shouldn't be an issue (also, note that the optimal solution under Nash-MTL for the two tasks case is equivalent to independently normalizing each gradient and summing with equal weights).
However, if you are solving a regression task, I suggest you normalize the target for each task if the scale of the target is large.

**Comment by yushengjiexy** [comment]
hi, i don't think this is a bug.Because GradNorm and PCGrad works.I think this is a limitation of the algorithm.
This paper may support my viewpoint:https://arxiv.org/pdf/2209.11379.pdf

**Comment by AvivNavon** [comment]
OK, good to hear.
I would still try to normalize each target vector.

**Comment by yushengjiexy** [comment]
thans
